orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/util/RegionCollector.java

```diff
@@ -0,0 +1,73 @@
+package com.netflix.spinnaker.orca.pipeline.util;
+
+import com.netflix.spinnaker.orca.pipeline.model.Stage;
+import org.jetbrains.annotations.NotNull;
+import org.springframework.stereotype.Component;
+
+import java.util.*;
+
+@Component
+/**
+ * Collects relevant regions from stages
+ *
+ * Used by e.g. FindImageFromClusterTask and BakeTask
+ */
+public class RegionCollector {
+  /**
+   * Traverses all descendant stages of a given stage looking for deploy/deploy canary stages
+   * When found, extracts the regions those stages deploy to.
+   *
+   * @param stage Stage for which to traverse the regions for
+   * @return union of all regions or an empty set
+   */
+  public @NotNull
+  Set<String> getRegionsFromDescendentStages(Stage stage) {
+    Set<String> deployRegions = new HashSet<>();
+    String stageCloudProviderType = stage.getContext().get("cloudProviderType") != null ?
```

Comment:
Might be a little tidier as: 

```
Map<String, Object> context = stage.getContext();
String cloudProvider = (String) context.getOrDefault("cloudProviderType", context.get("cloudProvider"))
```

Reply:
👍

Reply:
@ajordens : Btw, is it safe to assume that one of these keys (`cloudProviderType` or `cloudProvider`) will be set? I saw that `CloudProviderAware.getCloudProvider` always falls back on some default (`aws`) I should probably do the same in absence of strongly typed contexts?